ConcurrentModificationException in class VisitorPartialEvaluator method visitCtFor

Dear spoon developers,

I stumbled over an ConcurrentModificationException at VisitorPartialEvaluator.java:335 method visitCtFor were the code looks like this.

```java
                // Evaluate forInit
		List<CtStatement> lst = forLoop.getForInit();
		for (CtStatement s : lst) {
			CtStatement evaluateStatement = evaluate(s);
			if (evaluateStatement != null) {
				forLoop.addForInit(evaluateStatement);
			}
		}
```

If the `evaluate(s)` does return a instance, the `forLoop.addForInit` modifies the underlying list which result in a concurrent modification.
This happens if I try to use the template substitution where the CtStatement List which I like to print as template body does contain a for loop like this one

```java
    	for (int i = 0; i < 10; i++) {
			
	}
```

I'm using the spoon-core version 8.2.0
The full stack trace is

```text
Caused by: java.util.ConcurrentModificationException
	at java.util.ArrayList$Itr.checkForComodification(ArrayList.java:909)
	at java.util.ArrayList$Itr.next(ArrayList.java:859)
	at spoon.support.reflect.eval.VisitorPartialEvaluator.visitCtFor(VisitorPartialEvaluator.java:335)
	at spoon.support.reflect.code.CtForImpl.accept(CtForImpl.java:40)
	at spoon.support.reflect.eval.VisitorPartialEvaluator.evaluate(VisitorPartialEvaluator.java:95)
	at spoon.support.reflect.eval.VisitorPartialEvaluator.visitCtBlock(VisitorPartialEvaluator.java:246)
	at spoon.support.reflect.code.CtBlockImpl.accept(CtBlockImpl.java:57)
	at spoon.support.reflect.eval.VisitorPartialEvaluator.evaluate(VisitorPartialEvaluator.java:95)
	at spoon.support.reflect.code.CtCodeElementImpl.partiallyEvaluate(CtCodeElementImpl.java:24)
	at spoon.pattern.internal.DefaultGenerator.lambda$generateTargets$0(DefaultGenerator.java:88)
	at spoon.pattern.internal.ResultHolder$Single.mapEachResult(ResultHolder.java:88)
	at spoon.pattern.internal.DefaultGenerator.generateTargets(DefaultGenerator.java:84)
	at spoon.pattern.internal.DefaultGenerator.generateSingleTarget(DefaultGenerator.java:55)
	at spoon.pattern.internal.node.ElementNode.generateSingleNodeAttributes(ElementNode.java:278)
	at spoon.pattern.internal.node.ElementNode.generateTargets(ElementNode.java:268)
	at spoon.pattern.internal.DefaultGenerator.generateTargets(DefaultGenerator.java:81)
	at spoon.pattern.internal.node.ListOfNodes.generateTargets(ListOfNodes.java:41)
	at spoon.pattern.internal.DefaultGenerator.generateTargets(DefaultGenerator.java:81)
	at spoon.pattern.internal.DefaultGenerator.generateTargets(DefaultGenerator.java:69)
	at spoon.pattern.internal.node.ElementNode.generateSingleNodeAttributes(ElementNode.java:281)
	at spoon.pattern.internal.node.ElementNode.generateTargets(ElementNode.java:268)
	at spoon.pattern.internal.DefaultGenerator.generateTargets(DefaultGenerator.java:81)
	at spoon.pattern.internal.DefaultGenerator.generateSingleTarget(DefaultGenerator.java:55)
	at spoon.pattern.internal.node.ElementNode.generateSingleNodeAttributes(ElementNode.java:278)
	at spoon.pattern.internal.node.ElementNode.generateTargets(ElementNode.java:268)
	at spoon.pattern.internal.DefaultGenerator.generateTargets(DefaultGenerator.java:81)
	at spoon.pattern.internal.node.ListOfNodes.generateTargets(ListOfNodes.java:41)
	at spoon.pattern.internal.DefaultGenerator.generateTargets(DefaultGenerator.java:81)
	at spoon.pattern.internal.DefaultGenerator.generateTargets(DefaultGenerator.java:69)
	at spoon.pattern.internal.DefaultGenerator.generate(DefaultGenerator.java:224)
	at spoon.template.TemplateBuilder.substituteSingle(TemplateBuilder.java:162)
	at spoon.template.Substitution.substitute(Substitution.java:524)
	at spoon.template.Substitution.insertMethod(Substitution.java:396)
	at spoon.template.Substitution.insertAllMethods(Substitution.java:219)
	at spoon.template.Substitution.insertAllMethods(Substitution.java:183)
	at spoon.template.Substitution.insertAll(Substitution.java:64)
	at abc.ccf.refactoring.impl.RClassTemplateImpl.createClass(RClassTemplateImpl.java:63)
	at SpringControllerRefactoring.extractController(SpringControllerRefactoring.java:105)
	at java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:184)
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
	at java.util.TreeMap$KeySpliterator.forEachRemaining(TreeMap.java:2746)
	at java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:580)
	at java.util.stream.ReferencePipeline$7$1.accept(ReferencePipeline.java:270)
	at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:175)
	at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1382)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
	at java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151)
	at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:418)
	at abc.ccf.refactoring.impl.RDDLImpl.executeOnMethods(RDDLImpl.java:65)
	at SpringControllerRefactoring.Extract_identified_code_and_extract_to_spring_controller(SpringControllerRefactoring.java:67)
	at SpringControllerRefactoring.execute(SpringControllerRefactoring.java:57)
	at SpringControllerRefactoring.run(SpringControllerRefactoring.java:43)
	at SpringControllerRefactoring.main(SpringControllerRefactoring.java:32)
	... 8 more
```

Please let me know if additional information is required.
I'm looking forward to have this small bug fixed

Best wishes to the lovely city of Lille
-Matthias

--- a/src/main/java/spoon/support/reflect/eval/VisitorPartialEvaluator.java
+++ b/src/main/java/spoon/support/reflect/eval/VisitorPartialEvaluator.java
@@ -47,6 +47,7 @@ import spoon.reflect.reference.CtTypeReference;
 import spoon.reflect.visitor.CtScanner;
 import spoon.support.util.RtHelper;
 
+import java.util.ArrayList;
 import java.util.List;
 
 /**
@@ -332,7 +333,7 @@ public class VisitorPartialEvaluator extends CtScanner implements PartialEvaluat
 
 		// Evaluate forInit
 		List<CtStatement> lst = forLoop.getForInit();
-		for (CtStatement s : lst) {
+		for (CtStatement s : new ArrayList<>(lst)) { // copying the list to avoid ConcurrentModificationException
 			CtStatement evaluateStatement = evaluate(s);
 			if (evaluateStatement != null) {
 				forLoop.addForInit(evaluateStatement);
@@ -344,7 +345,7 @@ public class VisitorPartialEvaluator extends CtScanner implements PartialEvaluat
 
 		// Evaluate forUpdate
 		lst = forLoop.getForUpdate();
-		for (CtStatement s : lst) {
+		for (CtStatement s : new ArrayList<>(lst)) { // copying the list to avoid ConcurrentModificationException
 			CtStatement evaluateStatement = evaluate(s);
 			if (evaluateStatement != null) {
 				forLoop.addForUpdate(evaluateStatement);

